builtin/logical/pki/path_config_ca.go

```diff
@@ -2,33 +2,461 @@ package pki
+		Type: framework.TypeString,
+		Description: `Must be "internal" or "exported".
+If set to "exported", the generated private
+key will be returned. This is your *only*
+chance to retrieve the private key!`,
+	}
+
+	ret.Fields["key_bits"] = &framework.FieldSchema{
+		Type:    framework.TypeInt,
+		Default: 2048,
+		Description: `The number of bits to use. You will almost
+certainly want to change this if you adjust
+the key_type.`,
+	}
+
+	return ret
+}
+
+func pathSignIntermediateCA(b *backend) *framework.Path {
+	ret := &framework.Path{
+		Pattern: "config/ca/sign",
+
+		Fields: rootAndSignSchema,
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.WriteOperation: b.pathCASignWrite,
 		},
 
-		HelpSynopsis:    pathConfigCAHelpSyn,
-		HelpDescription: pathConfigCAHelpDesc,
+		HelpSynopsis:    pathConfigCASignHelpSyn,
+		HelpDescription: pathConfigCASignHelpDesc,
 	}
+
+	ret.Fields["pki_address"] = &framework.FieldSchema{
+		Type: framework.TypeString,
+		Description: `The base URL of the *destination*
+PKI mount, e.g.
+"https://vault.example.com/v1/intermediate_pki".
+For HA setups, the given host name
+should be the address that can always
+be used to contact the leader, as this is
+is used for generating the CA/CRL URLs in
+the certificate.`,
+	}
+
+	ret.Fields["csr"] = &framework.FieldSchema{
+		Type:        framework.TypeString,
+		Description: `PEM-format CSR to be signed.`,
+	}
+
+	return ret
 }
 
-func (b *backend) pathCAWrite(
+func (b *backend) pathCAGenerateRoot(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	var err error
+	exported := data.Get("exported").(string)
+	switch exported {
+	case "exported":
+	case "internal":
+	default:
+		return logical.ErrorResponse(fmt.Sprintf(
+			"The \"exported\" path parameter must be \"internal\" or \"exported\"")), nil
+	}
+
+	pkiAddress := strings.ToLower(data.Get("pki_address").(string))
+	switch {
+	case len(pkiAddress) == 0:
+		return logical.ErrorResponse(fmt.Sprintf(
+			"\"pki_address\" cannot be empty")), nil
+	case !strings.HasPrefix(pkiAddress, "http"):
+		return logical.ErrorResponse(fmt.Sprintf(
+			"\"pki_address\" must be a URL")), nil
+	case !strings.Contains(pkiAddress, "/v1/"):
+		return logical.ErrorResponse(fmt.Sprintf(
+			"\"pki_address\" needs to be the path to the PKI mount, not the base Vault path")), nil
+	case !strings.Contains(pkiAddress, "/v1/"+req.MountPoint[:len(req.MountPoint)-1]):
+		return logical.ErrorResponse(fmt.Sprintf(
+			"\"pki_address\" needs to be the path to this mount")), nil
+	}
+	if strings.HasSuffix(pkiAddress, "/") {
+		pkiAddress = pkiAddress[:len(pkiAddress)-1]
+	}
+
+	role := &roleEntry{
+		TTL:              data.Get("ttl").(string),
+		KeyType:          "rsa",
+		KeyBits:          data.Get("key_bits").(int),
+		AllowLocalhost:   true,
+		AllowAnyName:     true,
+		EnforceHostnames: false,
+	}
+
+	switch role.KeyBits {
+	case 0:
+		role.KeyBits = 2048
```

Comment:
It's a guard against them putting in 0 manually, but I guess that's not really necessary...we should just error out on that instead.